### PR TITLE
Remix properties (DONT MERGE)

### PR DESCRIPTION
**Pull request by ThatZeoMan**
apparently this lets people work on my project at the same time as me. Yippee!

### CODE REVIEW

File: Assets/Scripts/LevelEditor/Properties/PropertiesDialog.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    public class PropertiesDialog : MonoBehaviour
+    {
+        [SerializeField] private GameObject propertiesMenu;
+        //this may all be moved to a different script in the future
+
+        private void Start() {}
+
+        public void SwitchPropertiesDialog()
+        {
+            if(propertiesMenu.activeSelf) {
```
Comment (huantianad) [inline Assets/Scripts/LevelEditor/Properties/PropertiesDialog.cs:19]:
```suggestion
            if (propertiesMenu.activeSelf) {
```

File: Assets/Scripts/LevelEditor/Editor.cs [hunk 2]
```diff
@@ -352,6 +356,8 @@ private void SaveRemixFile(string path)
                     var musicFile = archive.CreateEntry("song.ogg", System.IO.Compression.CompressionLevel.NoCompression);
                     using (var zipStream = musicFile.Open())
                         zipStream.Write(bytes, 0, bytes.Length);
+
+
```
Comment (huantianad) [inline Assets/Scripts/LevelEditor/Editor.cs:360]:
```suggestion
```

File: Assets/Scripts/LevelEditor/Editor.cs [hunk 3]
```diff
@@ -343,6 +343,10 @@ private void SaveRemixFile(string path)
                     using (var zipStream = levelFile.Open())
                         zipStream.Write(Encoding.UTF8.GetBytes(GetJson()), 0, Encoding.UTF8.GetBytes(GetJson()).Length);
 
+                    var PropertiesFile = archive.CreateEntry("properties.json", System.IO.Compression.CompressionLevel.NoCompression);
+                    using (var zipStream = PropertiesFile.Open())
```
Comment (huantianad) [inline Assets/Scripts/LevelEditor/Editor.cs:347]:
```suggestion
                    var propertiesFile = archive.CreateEntry("properties.json", System.IO.Compression.CompressionLevel.NoCompression);
                    using (var zipStream = propertiesFile.Open())
```

File: Assets/Scripts/LevelEditor/Editor.cs [hunk 4]
```diff
@@ -343,6 +343,10 @@ private void SaveRemixFile(string path)
...
+                        zipStream.Write(Encoding.UTF8.GetBytes(GetPropertiesJson()), 0, Encoding.UTF8.GetBytes(GetPropertiesJson()).Length);
```
Comment (huantianad) [inline Assets/Scripts/LevelEditor/Editor.cs:348]: Ideally we should only call `GetPropertiesJson()` and `GetJson()` once and store the value in a variable. That way we don't call a possibly expensive serialization twice.

Since this refactors old code, should probably be a separate commit and/or PR

File: Assets/Scripts/LevelEditor/Editor.cs [hunk 5]
```diff
@@ -369,6 +375,12 @@ public void LoadRemix(string json = "")
             Timeline.FitToSong();
         }
 
+        public void GetProperties(string json = "")
+        {
+            PropController.instance.LoadProperties(json);
+            Debug.Log("properties sent");
+        }
+
```
Comment (huantianad) [inline Assets/Scripts/LevelEditor/Editor.cs:383]: Perhaps we should copy the same naming scheme, i.e. use `LoadProperties` instead?

File: Assets/Scripts/LevelEditor/Editor.cs [hunk 6]
```diff
@@ -483,6 +509,13 @@ public string GetJson()
             return json;
         }
 
+        public string GetPropertiesJson()
+        {
+            string json = string.Empty;
+            json = JsonConvert.SerializeObject(PropController.instance.properties);
+            return json;
+        }
```
Comment (huantianad) [inline Assets/Scripts/LevelEditor/Editor.cs:517]: This can be simplified to:
```suggestion
        public string GetPropertiesJson() =>
            JsonConvert.SerializeObject(PropController.instance.properties);
```
Since this refactors old code, should probably be a separate commit and/or PR

File: Assets/Scripts/LevelEditor/Properties/PropertyDefn.cs [hunk 7]
```diff
@@ -0,0 +1,197 @@
+using System.Collections;
```
Comment (huantianad) [inline Assets/Scripts/LevelEditor/Properties/PropertyDefn.cs:1]: Is this code based off somewhere else? Could we reuse it directly instead of making it a new version of it?

**Review by huantianad — COMMENTED** [review]
Code looks pretty good overall! Didn't check Unity stuff but looks good.
The main question is if we want to make Properties separate or make it part of the Beatmap object, @minenice55 might have more thoughts.
Also I was wondering if perhaps Metadata would be a better name, or even better make it under the Beatmap namespace so it's `Beatmap.Metadata`.

File: Assets/Scripts/LevelEditor/Properties/PropController.cs [hunk 8]
```diff
@@ -44,14 +44,33 @@ private void Start()
...
         //this is just copied from the beatmap lol
         public string levelName = "test";
         public string levelCreator = "testCreator";
         public int Number;
 
+        public object this[string propertyName]
+        {
+            get
+            {
+                return typeof(Entity).GetField(propertyName).GetValue(this);
```
Comment (minenice55) [inline Assets/Scripts/LevelEditor/Properties/PropController.cs:59]: `typeof(Entity)` should be `typeof(Properties)` to refer to the correct class

File: Assets/Scripts/LevelEditor/Properties/PropController.cs [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+                try
+                {
+                    typeof(Properties).GetField(propertyName).SetValue(this, value);
+                }
+                catch (Exception ex)
+                {
+                    UnityEngine.Debug.LogError($"You probably misspelled a parameter, or defined the object type wrong. Exception log: {ex}");
+                }
+            }
+        }
```
Comment (huantianad) [inline Assets/Scripts/LevelEditor/Properties/PropController.cs:78]: this makes me sad inside, access the fields directly, eg `properties.levelName`

File: Assets/Scripts/LevelEditor/Properties/RemixParameterManager.cs [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+                prefab = DropdownP;
+            }
+            else if (objType == typeof(Color))
+            {
+                prefab = ColorP;
+            }
+            else if(objType == typeof(string))
+            {
+                prefab = StringP;
+            }
```
Comment (huantianad) [inline Assets/Scripts/LevelEditor/Properties/RemixParameterManager.cs:49]: Check out [switch expressions](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/switch-expression), it'll clean up the code a bit

**Comment by ThatZeoMan** [comment]
aight looks like mine is making the properties now. it was fun tho!
